### PR TITLE
drivers: gpio: Add Davinci gpio controller support

### PR DESCRIPTION
**Pull request by slpp95prashanth**
Davinci gpio controller support to add various soc gpio support (J721E, AM654).

TRM for J721e https://www.ti.com/lit/zip/spruil1
File: spruil1c.pdf
GPIO: section 12.1.2

BeagleBone AI_64 https://beagleboard.org/ai-64

### CODE REVIEW

File: drivers/gpio/gpio_davinci.c [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+	.port_clear_bits_raw = gpio_davinci_port_clear_bits_raw,
+	.port_toggle_bits = gpio_davinci_port_toggle_bits
+};
+
+static int gpio_davinci_init(const struct device *dev)
+{
+	return 0;
+}
+
+#define GPIO_J721E_INIT_FUNC(n)	\
```
Comment (DhruvaG2000) [inline drivers/gpio/gpio_davinci.c:149]: > GPIO_J721E_INIT_FUNC

Are we allowing device specific stuff in a driver? If it's not J721E specific change the naming.

File: drivers/gpio/gpio_davinci.c [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+	.port_get_raw = gpio_davinci_port_get_raw,
+	.port_set_masked_raw = gpio_davinci_port_set_masked_raw,
+	.port_set_bits_raw = gpio_davinci_port_set_bits_raw,
+	.port_clear_bits_raw = gpio_davinci_port_clear_bits_raw,
+	.port_toggle_bits = gpio_davinci_port_toggle_bits
+};
+
+static int gpio_davinci_init(const struct device *dev)
+{
+	return 0;
```
Comment (DhruvaG2000) [inline drivers/gpio/gpio_davinci.c:146]: you really don't want to do anything in `_init`?

File: drivers/gpio/gpio_davinci.c [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+	DEVICE_DT_INST_DEFINE(n,	\
+		&gpio_davinci_init,	\
+		NULL,	\
+		&gpio_davinci_##n##_data,	\
+		&gpio_davinci_##n##_config,	\
+		PRE_KERNEL_2,	\
+		CONFIG_GPIO_INIT_PRIORITY,	\
+		&gpio_davinci_driver_api);
+
+DT_INST_FOREACH_STATUS_OKAY(GPIO_J721E_INIT)
```
Comment (DhruvaG2000) [inline drivers/gpio/gpio_davinci.c:185]: Again, why `GPIO_J721E_INIT` ? Is it something specific to J721E? If yes then does it really belong here in the driver or somewhere more platform specific?

File: dts/bindings/gpio/ti,am654-gpio-nexus.yaml [hunk 4]
```diff
@@ -0,0 +1,16 @@
+# Copyright (c) 2021, Yonatan Schachter
+# SPDX-License-Identifier: Apache-2.0
+
+description: GPIO controller for Davinci and keystone devices
+
+compatible: "ti,j721e-gpio-nexus"
```
Comment (DhruvaG2000) [inline dts/bindings/gpio/ti,am654-gpio-nexus.yaml:6]: The Linux kernel already has:
```
.compatible = "ti,am654-gpio"
```

Why not reuse?

**Review by DhruvaG2000 — CHANGES_REQUESTED** [review]
Have left inline comments

File: drivers/gpio/gpio_davinci.c (hunk 3 above)
Comment (slpp95prashanth) [reply drivers/gpio/gpio_davinci.c:185]: updating to DAVINCI.

File: dts/bindings/gpio/ti,am654-gpio-nexus.yaml (hunk 4 above)
Comment (slpp95prashanth) [reply dts/bindings/gpio/ti,am654-gpio-nexus.yaml:6]: ti,j721e-gpio-nexus is zephyr specific for having sub nodes for gpio in device tree, as the number of pins for a gpio controller is 32.

The compatible for the node is ti,am654-gpio-nexus and submodes are ti,am654-gpio.

Comment (slpp95prashanth) [reply dts/bindings/gpio/ti,am654-gpio-nexus.yaml:6]: ti,j721e-gpio-nexus is zephyr specific for having sub nodes for gpio in device tree, as the number of pins for a gpio controller is 32.

The compatible for the node is ti,am654-gpio-nexus and submodes are ti,am654-gpio.

File: drivers/gpio/gpio_davinci.c (hunk 2 above)
Comment (slpp95prashanth) [reply drivers/gpio/gpio_davinci.c:146]: In the init we should register interrupts for gpio controllers, In J721e, gpio interrupts are through interrupt router to configure that we need ti,sci support, which is not added in zephyr.

**Review by dnltz — COMMENTED** [review]
Left some inline comments too.

File: drivers/gpio/gpio_davinci.c [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+
+	} else {
+		regs->dir |= (BIT(pin));
+	}
+
+	return 0;
+}
+
+static int gpio_davinci_port_get_raw(const struct device *dev,
+									gpio_port_value_t *value)
```
Comment (dnltz) [inline drivers/gpio/gpio_davinci.c:82]: Looks like the way you broke the arguments made it actually worser and will probably exceed 80 characters. These arguments should be placed close to the opening bracket / first argument.
```suggestion
static int gpio_davinci_port_get_raw(const struct device *dev,
					gpio_port_value_t *value)
```

**Review by dnltz — CHANGES_REQUESTED** [review]
Good job! I only see one issue with how you broke the arguments.

File: drivers/gpio/gpio_davinci.c (hunk 5 above)
Comment (slpp95prashanth) [reply drivers/gpio/gpio_davinci.c:82]: @dnltz I will align the arguments to the opening bracket (my vim editor has tabspace set to 2 spaces, so the alignment looked fine in my editor, I will change it).

Comment (kartben) [reply drivers/gpio/gpio_davinci.c:82]: FWIW tab space is 8 for Zephyr, and max line length is 100 characters (not 80).

**Review by dnltz — APPROVED** [review]
Looks good to me :)

**Comment by DhruvaG2000** [comment]
@slpp95prashanth  any reason to still keep this a draft PR? Please mark as ready if you finally think it's in a good state to review and merge. (unless there's some "depends on" PR that needs to go in before this). If so then please specify which PR

File: drivers/gpio/gpio_davinci.c [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+#define DT_DRV_COMPAT ti_davinci_gpio
+
+#include <errno.h>
+#include <zephyr/kernel.h>
+#include <zephyr/device.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/arch/common/sys_bitops.h>
+#include <zephyr/devicetree.h>
+
+#include <zephyr/drivers/gpio/gpio_utils.h>
```
Comment (DhruvaG2000) [inline drivers/gpio/gpio_davinci.c:17]: Have you sorted all the includes alphabetically?

File: dts/bindings/gpio/ti,davinci-gpio.yaml [hunk 7]
```diff
@@ -0,0 +1,21 @@
+# Copyright (C) 2023 BeagleBoard.org Foundation
+# Copyright (C) 2023 S Prashanth
+#
+# SPDX-License-Identifier: Apache-2.0
+
+description: GPIO controller for Davinci and keystone devices.
```
Comment (DhruvaG2000) [inline dts/bindings/gpio/ti,davinci-gpio.yaml:6]: Capital K in keystone?

File: drivers/gpio/gpio_davinci.c [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+					gpio_port_value_t *value)
+{
+	volatile struct gpio_davinci_regs *regs = DEV_GPIO_CFG_BASE(dev);
+
+	*value = regs->in_data;
+
+	return 0;
+}
+
+static int gpio_davinci_port_set_masked_raw(const struct device *dev,
```
Comment (DhruvaG2000) [inline drivers/gpio/gpio_davinci.c:91]: Each of the functions in this driver could use some documentation, for example
https://elixir.bootlin.com/linux/v6.5-rc4/source/drivers/gpio/gpio-davinci.c#L130

See how linux documents most of these gpio funcs explaining briefly what they're doing.

File: drivers/gpio/gpio_davinci.c [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+	return 0;
+}
+
+static const struct gpio_driver_api gpio_davinci_driver_api = {
+	.pin_configure = gpio_davinci_configure,
+	.port_get_raw = gpio_davinci_port_get_raw,
+	.port_set_masked_raw = gpio_davinci_port_set_masked_raw,
+	.port_set_bits_raw = gpio_davinci_port_set_bits_raw,
+	.port_clear_bits_raw = gpio_davinci_port_clear_bits_raw,
+	.port_toggle_bits = gpio_davinci_port_toggle_bits
```
Comment (DhruvaG2000) [inline drivers/gpio/gpio_davinci.c:137]: Minor nit,
Can we not reuse the kernel naming convention just for maintaining similarity?
Ie. davinci_gpio_xxx
https://elixir.bootlin.com/linux/v6.5-rc4/source/drivers/gpio/gpio-davinci.c#L145

**Review by DhruvaG2000 — CHANGES_REQUESTED** [review]
Left some comments, please take a look

File: drivers/gpio/gpio_davinci.c (hunk 9 above)
Comment (slpp95prashanth) [reply drivers/gpio/gpio_davinci.c:137]: In zephyr, all gpio driver functions are defined as gpio_<module_name>.

File: drivers/gpio/gpio_davinci.c (hunk 8 above)
Comment (slpp95prashanth) [reply drivers/gpio/gpio_davinci.c:91]: Discussed in discord https://discord.com/channels/1108795636956024986/1109167658131275886/1141376293850595378

**Review by DhruvaG2000 — APPROVED** [review]
LGTM. thanks for addressing all my concerns!

File: drivers/gpio/gpio_davinci.c [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+#include <zephyr/device.h>
+#include <zephyr/devicetree.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/drivers/gpio/gpio_utils.h>
+#include <zephyr/kernel.h>
+
+/* Helper Macros for GPIO */
+#define DEV_GPIO_CFG(dev)	\
+		((const struct gpio_davinci_config * const)((dev)->config))
+#define DEV_GPIO_CFG_BASE(dev) \
```
Comment (vaishnavachath) [inline drivers/gpio/gpio_davinci.c:22]: @slpp95prashanth can we use DEVICE_MMIO_GET()/DEVICE_MMIO_NAMED_GET()  to get the base address? this driver should be generic and should work with other K3 platforms as well.

**Review by vaishnavachath — CHANGES_REQUESTED** [review]

File: drivers/gpio/gpio_davinci.c [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+	DEVICE_DT_INST_DEFINE(n,	\
+		&gpio_davinci_init,	\
+		NULL,	\
+		&gpio_davinci_##n##_data,	\
+		&gpio_davinci_##n##_config,	\
+		PRE_KERNEL_2,	\
+		CONFIG_GPIO_INIT_PRIORITY,	\
+		&gpio_davinci_driver_api);
+
+DT_INST_FOREACH_STATUS_OKAY(GPIO_DAVINCI_INIT)
```
[the remark this answers is not in the document]
Comment (slpp95prashanth) [reply drivers/gpio/gpio_davinci.c:189]: The power and clocks are enabled by the west (Linux) same as for all the modules, as we do not have sci support.

I tested the blinky example on bbai 64 board.

**Review by vaishnavachath — DISMISSED** [review]
Looks Good, this should ideally be merged only after https://github.com/zephyrproject-rtos/zephyr/pull/59191 is merged, if someone tests and confirms functionality on AM62 A53/M4 target that would be great.

**Comment by dnltz** [comment]
> Looks Good, this should ideally be merged only after #59191 is merged, if someone tests and confirms functionality on AM62 A53/M4 target that would be great.

I can do that today/tomorrow!

**Review by aasinclair — DISMISSED** [review]
Looks good.
Some of the whitespace looks a little strange, for future contributions you may wish to consider using clang-format to autoformat the code, rather than aligning things manually.

**Review by aasinclair — APPROVED** [review]

**Review by vaishnavachath — APPROVED** [review]

File: drivers/gpio/gpio_davinci.c [hunk 12]
```diff
@@ -0,0 +1,199 @@
...
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/drivers/gpio/gpio_utils.h>
+#include <zephyr/init.h>
+#include <zephyr/kernel.h>
+#include <zephyr/sys/sys_io.h>
+
+/* Helper Macros for GPIO */
+#define DEV_CFG(dev) \
+		((const struct gpio_davinci_config * const)((dev)->config))
+#define DEV_DATA(dev) ((struct gpio_davinci_data *)(dev)->data)
```
Comment (gmarull) [inline drivers/gpio/gpio_davinci.c:24]: this was removed from all drivers long time ago, please, drop such accessors.

**Review by gmarull — DISMISSED** [review]

File: drivers/gpio/gpio_davinci.c (hunk 12 above)
Comment (gmarull) [reply drivers/gpio/gpio_davinci.c:24]: actually I was wrong, this uses the MMIO API, so it is still required

**Pull request merged**